nucypher/crypto/keystore.py

```diff
@@ -0,0 +1,415 @@
+    return __private
+
+
+def _assemble_keystore(encrypted_secret: bytes, salt: bytes) -> Dict[str, Union[str, bytes]]:
+    encoded_key_data = {
+        'version': '2.0',
+        'created': str(time.time()),
+        'key': encrypted_secret,
+        'salt': salt
+    }
+    return encoded_key_data
+
+
+def _read_keystore(path: Path, deserializer: Callable) -> Dict[str, Union[str, bytes]]:
+    """Parses a keyfile and return decoded, deserialized key data."""
+    with open(path, 'rb') as keyfile:
+        key_data = keyfile.read()
+        if deserializer:
+            key_data = deserializer(key_data)
+    return key_data
+
+
+def _write_keystore(path: Path, payload: Dict[str, bytes], serializer: Callable) -> Path:
+    """
+    Creates a permissioned keyfile and save it to the local filesystem.
+    The file must be created in this call, and will fail if the path exists.
+    Returns the filepath string used to write the keyfile.
+
+    Note: getting and setting the umask is not thread-safe!
+
+    See linux open docs: http://man7.org/linux/man-pages/man2/open.2.html
+    ---------------------------------------------------------------------
+    O_CREAT - If pathname does not exist, create it as a regular file.
+
+
+    O_EXCL - Ensure that this call creates the file: if this flag is
+             specified in conjunction with O_CREAT, and pathname already
+             exists, then open() fails with the error EEXIST.
+    ---------------------------------------------------------------------
+    """
+
+    if path.exists():
+        raise Keystore.Exists(f"Private keyfile {path} already exists.")
+    try:
+        keyfile_descriptor = os.open(path, flags=__PRIVATE_FLAGS, mode=__PRIVATE_MODE)
+    finally:
+        os.umask(0)  # Set the umask to 0 after opening
+    if serializer:
+        payload = serializer(payload)
+    with os.fdopen(keyfile_descriptor, 'wb') as keyfile:
+        keyfile.write(payload)
+    return path
+
+
+def _serialize_keystore(payload: Dict) -> bytes:
+    for field in ('key', 'salt'):
+        payload[field] = bytes(payload[field]).hex()
+    try:
+        metadata = json.dumps(payload, indent=4)
+    except JSONDecodeError:
+        raise Keystore.Invalid("Invalid or corrupted key data")
+    return bytes(metadata, encoding=FILE_ENCODING)
+
+
+def _deserialize_keystore(payload: bytes):
+    payload = payload.decode(encoding=FILE_ENCODING)
+    try:
+        payload = json.loads(payload)
+    except JSONDecodeError:
+        raise Keystore.Invalid("Invalid or corrupted key data")
+    for field in ('key', 'salt'):
+        payload[field] = bytes.fromhex(payload[field])
+    return payload
+
+
+def generate_keystore_filepath(parent: Path, id: str) -> Path:
+    utc_nowish = int(time.time())  # epoch
+    path = parent / f'{utc_nowish}-{id}.priv'
+    return path
+
+
+def validate_keystore_password(password: str) -> List:
+    """
+    NOTICE: Do not raise inside this function.
```

Comment:
This is a fairly dated comment now, but it's still meant to describe two cases:

- (Yes) Logging of the local scope's context to sentry, which includes the password.
- Usage as a `click` interactive prompt validator which requires a boolean response.

Reply:
> Logging of the local scope's context to sentry, which includes the password.

Would it make sense to pass around the password in a container class with redefined `str`/`repr`, unpacking only right before usage? Or is it an overcomplication?

Reply:
> Would it make sense to pass around the password in a container class with redefined

Hmm not a bad idea, I wonder if this can effectively prevent unintentional transmission...  Sounds like something that belongs in the `passwords.py` module from your recent PR, but perhaps in a follow up.  I'll open an issue about this after the review process advances a bit.